test: execute testing

agureev opened the pull request:
Based on #244 by @murisi

Changelog:

Simplifications:

- Removal of structs in favor of explicit fuzzing arguments
- Minimize assumes and replace them with nonced computations based on fuzzing arguments
- Abstract bounding of fuzzing variables
- Make Mock just a test again

Improvements:

- Generalize ResourceCount reverts: tests not accepts more inputs, not only less

General comments:

- Made the output test into a unit one as there is no proper way to fuzz it currently.
- Certain mutations are assumed to fail early so are not properly generated to replicate the original constructions.

heueristik commented on contracts/test/ProtocolAdapterMock.t.sol:
```diff
@@ -238,6 +241,315 @@ contract ProtocolAdapterMockVerifierTest is Test {
+    }
+
+    /// @notice Test that transactions with a missing logic verifier input fail.
+    function testFuzz_execute_incorrect_logic_verifier_count_fail(
```
```suggestion
    function testFuzz_execute_reverts_on_logic_and_compliance_verifier_tag_count_mismatch(
```
I am not happy with this and the previous name yet

heueristik commented on contracts/test/ProtocolAdapterMock.t.sol:
```diff
@@ -238,6 +241,315 @@ contract ProtocolAdapterMockVerifierTest is Test {
+        _mockPa.execute(txn);
+    }
+
+    /// @notice Test that transactions with mismatching nullfifier logic references fail.
+    function testFuzz_execute_mismatching_nullifier_logic_refs_fail(
```
```suggestion
    function testFuzz_execute_reverts_on_compliance_and_logic_verifier_logic_reference_mismatch_for_consumed_resource(
```

heueristik commented on contracts/test/ProtocolAdapterMock.t.sol:
```diff
@@ -238,6 +241,315 @@ contract ProtocolAdapterMockVerifierTest is Test {
+    }
+
+    /// @notice Test that transactions with unknown selectors fail.
+    function testFuzz_execute_unknown_selector_fails(
```
```suggestion
    function testFuzz_execute_reverts_if_proofs_start_with_an_unknown_verifier_selector(
```

heueristik commented on contracts/test/ProtocolAdapterMock.t.sol:
```diff
@@ -238,6 +241,315 @@ contract ProtocolAdapterMockVerifierTest is Test {
+    ) public {
+        // Ensure that the first 4 bytes of the proof are invalid.
+        // The mock router only accepts one selector we deploy with.
+        vm.assume(bytes4(fakeProof) != bytes4(0xFFFFFFFF));
```
```suggestion
        vm.assume(bytes4(fakeProof) != _mockVerifier.SELECTOR());
```

murisi reviewed (dismissed):
Thanks for working on these changes, they look good to me. Just some comments:
* These updates would prevent these tests from being used with the real protocol adapter/prover (as was possible with #244). Is coverage of the real protocol adapter a goal?
* The fusing of transaction generation with the testing logic, and increase in assumptions the test logic makes about the generated transactions might make it harder to expand the transaction generation code to, say, generate compliance units in a different order from logic units, or generate differing numbers of compliance units for each action. Do we ultimately want to broadly fuzz the protocol adapter to find potential abnormalities?
* Maybe the implicit assumptions being made about the protocol adapter/prover and the generated transactions could be noted? This would make it easier to debug these tests if a future protocol adapter breaks them (without necessarily breaking the specifications), say, by doing its checks in a slightly different order.

In conclusion, if we do not harbour any future ambitions of generating a broader range of test transactions or also testing the real protocol adapter with the same logic, then these readability and `vm.assume` reducing optimizations are fine. Thanks again!

agureev commented on contracts/test/ProtocolAdapterMock.t.sol:
```diff
@@ -238,6 +241,315 @@ contract ProtocolAdapterMockVerifierTest is Test {
+        );
+
+        // Replace the nullifier corresponding to the selected compliance unit with a fake one.
+        txn.actions[actionIndex].logicVerifierInputs[complianceIndex * 2].tag = fakeTag;
```
Yes, that is a good point. Here we are relying hat there is a "canonical" ordering that actually both us and arm-risc0 uses to generate examples.

But we should note this explicitly.

agureev replied:
Assumption removed. See latest commit.

agureev commented on contracts/test/ProtocolAdapterMock.t.sol:
```diff
@@ -238,6 +241,315 @@ contract ProtocolAdapterMockVerifierTest is Test {
+    }
+
+    /// @notice Test that transactions with mismatching commitment logic references fail.
+    function testFuzz_execute_mismatching_commitment_logic_refs_fail(
```
I'll think about whether there is a better way of doing this, yeah.

agureev replied:
Ditto

agureev commented on contracts/test/ProtocolAdapterMock.t.sol:
```diff
@@ -238,6 +241,309 @@ contract ProtocolAdapterMockVerifierTest is Test {
+        _mockPa.execute(txn);
+    }
+
+    function testFuzz_execute_reverts_if_commitment_from_compliance_inputs_cannot_be_found_in_logic_inputs(
```
Ditto

agureev commented on contracts/test/ProtocolAdapterMock.t.sol:
```diff
@@ -238,6 +241,309 @@ contract ProtocolAdapterMockVerifierTest is Test {
+        // Generate a fake logic using a nonce.
+        bytes32 fakeLogic = SHA256.hash(consumed.logicRef, nonce);
+        // Replace the original logic.
+        txn.actions[actionIndex].logicVerifierInputs[(complianceIndex * 2) + 1].verifyingKey = fakeLogic;
```
Yes

heueristik commented on contracts/test/ProtocolAdapterMock.t.sol:
```diff
@@ -238,6 +241,309 @@ contract ProtocolAdapterMockVerifierTest is Test {
+        );
+
+        // Replace the nullifier corresponding to the selected compliance unit with a fake one.
+        txn.actions[actionIndex].logicVerifierInputs[complianceIndex * 2].tag = fakeTag;
```
```suggestion
        txn.actions[actionIndex].logicVerifierInputs[complianceIndex * 2].tag = bytes32(type(uint256).max);
```

agureev commented on contracts/test/ProtocolAdapterMock.t.sol:
```diff
@@ -238,6 +241,315 @@ contract ProtocolAdapterMockVerifierTest is Test {
+
+        (Transaction memory txn,) = vm.transaction({mockVerifier: _mockVerifier, nonce: 0, configs: configs});
+
+        // Generate a different tag with the nonce.
```
Given that in the transaction generation our tags are generated via `sha256` of the encoded bytes, the fact that the fake tag coincides should be equivalent to hash-breaking

Hence I will let this stay

However this is certainly important to address in the comment, thank you for pointing it out.

agureev commented on contracts/test/ProtocolAdapterMock.t.sol:
```diff
@@ -238,6 +241,315 @@ contract ProtocolAdapterMockVerifierTest is Test {
+        bytes memory fakeProof
+    ) public {
+        // Ensure that the proof does not have enough bytes to contain the selector.
+        vm.assume(fakeProof.length < 4);
```
Removed the test based on suggestions.

agureev commented:
Changelog:

- Comment regarding `TagMismatch` check added
- Added `fakeTag` generation comment justifying tag uniqueness
- Added `getExistingTagIndex` function in `TxGen`
- Abstracted fuzzing examples using tag placement now using general tag placement

Comments:

- Removed short proof tests per @mariari comments
- Based on private conversations, the comment of @murisi regarding examples used for proper PA resolved. All examples use actual PA code and do not assume anything but the canonical mocked verifier.

agureev commented on contracts/test/ProtocolAdapterMock.t.sol:
```diff
@@ -238,6 +241,282 @@ contract ProtocolAdapterMockVerifierTest is Test {
+        bytes32 tag = txn.actions[actionIndex].complianceVerifierInputs[complianceIndex].instance.consumed.nullifier;
+        uint256 tagIndex = TxGen.getExistingTagIndex(txn.actions[actionIndex], tag);
+
+        // Generate a different tag with the nonce.
+        // We assume that the tags are generated using sha256. Hence the tag is different modulo hash-breaking.
+        bytes32 fakeTag = SHA256.hash(tag, nonce);
```
Hm, I guess so, however I don't think I am using "modulo" here in any formal sense. But I guess we can remove it. However yes, without removal of the tag generation.

murisi reviewed (approved):
Looks fine to me, thanks!
